prov/sockets/src/sock_progress.c

```diff
@@ -2629,10 +2630,12 @@ static void sock_pe_poll(struct sock_pe *pe)
 	if (pe->waittime && ((fi_gettime_ms() - pe->waittime) < sock_pe_waittime))
 		return;
 
-	if (dlist_empty(&pe->tx_list) && dlist_empty(&pe->rx_list))
+	pthread_mutex_lock(&pe->list_lock);
+	if (dlist_empty(&pe->tx_list) && dlist_empty(&pe->rx_list)) {
+		pthread_mutex_unlock(&pe->list_lock);
 		return;
+	}
 
-	pthread_mutex_lock(&pe->list_lock);
 	if (!dlist_empty(&pe->tx_list)) {
```

Comment:
Looking closer at the source, here's what I think makes sense to do as a first pass:

1. Split sock_pe_poll() into 2 calls: sock_pe_wait_ok() and sock_pe_wait().
2. sock_pe_wait() would be from 2666 and down - starting at the call to sock_epoll_wait().
3. sock_pe_wait_ok() is line 2666 and up.
4. Remove the mutex from  sock_pe_wait_ok().
5. Change the loop in sock_pe_progress_thread() to:
```c
while (*((volatile int *)&pe->do_progress)) {
    pthread_mutex_lock(&pe->list_lock);
    if (pe->domain->progress_mode == FI_PROGRESS_AUTO) { /* consider moving this check into wait_ok */
        if (sock_pe_wait_ok(pe)) {
            pthread_mutex_unlock(&pe->list_lock);
            sock_pe_wait(pe);
            pthread_mutex_lock(&pe->list_lock);
        }

    if (!dlist_empty(&pe->tx_list)) {
        ...
```
This will avoid acquiring/releasing the mutex multiple times inside the progress loop.  There may be a way to merge sock_pe_wait_ok() with sock_pe_progress_thread() -- they have very similar code -- but that would be for another day.